.github/workflows/bin-image.yml

```
@@ -0,0 +1,184 @@
+      -
+        name: Login to Docker Hub
+        if: github.event_name != 'pull_request' && github.repository == 'moby/vpnkit'
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      -
+        name: Build
+        id: bake
+        uses: docker/bake-action@v6
+        with:
+          source: .
+          files: |
+            ./docker-bake.hcl
+            /tmp/bake-meta.json
+          targets: bin-image
+          set: |
+            *.platform=${{ matrix.platform }}
+            *.output=type=image,name=${{ env.REPO_SLUG }},push-by-digest=true,name-canonical=true,push=${{ github.event_name != 'pull_request' && github.repository == 'moby/vpnkit' }}
+            *.tags=
+      -
+        name: Export digest
+        if: github.event_name != 'pull_request' && github.repository == 'moby/vpnkit'
+        run: |
+          mkdir -p /tmp/digests
+          digest="${{ fromJSON(steps.bake.outputs.metadata)['bin-image']['containerimage.digest'] }}"
+          touch "/tmp/digests/${digest#sha256:}"
+      -
+        name: Upload digest
+        if: github.event_name != 'pull_request' && github.repository == 'moby/vpnkit'
+        uses: actions/upload-artifact@v4
+        with:
+          name: digests-${{ env.PLATFORM_PAIR }}
+          path: /tmp/digests/*
+          if-no-files-found: error
+          retention-days: 1
+
+  merge:
+    runs-on: ubuntu-24.04
+    timeout-minutes: 20
+    needs:
+      - build
+    if: always() && !contains(needs.*.result, 'failure') && !contains(needs.*.result, 'cancelled') && github.event_name != 'pull_request' && github.repository == 'moby/vpnkit'
+    steps:
+      -
+        name: Download meta bake definition
+        uses: actions/download-artifact@v4
+        with:
+          name: bake-meta
+          path: /tmp
+      -
+        name: Download digests
+        uses: actions/download-artifact@v4
+        with:
+          path: /tmp/digests
+          pattern: digests-*
+          merge-multiple: true
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v3
+        with:
+          buildkitd-flags: --debug
+      -
+        name: Login to Docker Hub
+        uses: docker/login-action@v3
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      -
+        name: Create manifest list and push
+        working-directory: /tmp/digests
+        run: |
+          set -x
+          docker buildx imagetools create $(jq -cr '.target."docker-metadata-action".tags | map("-t " + .) | join(" ")' /tmp/bake-meta.json) \
+            $(printf '${{ env.REPO_SLUG }}@sha256:%s ' *)
+      -
+        name: Inspect image
+        run: |
+          set -x
+          docker buildx imagetools inspect ${{ env.REPO_SLUG }}:$(jq -cr '.target."docker-metadata-action".args.DOCKER_META_VERSION' /tmp/bake-meta.json)
```

I think we could make use of this reusable workflow https://github.com/crazy-max/.github?tab=readme-ov-file#build-distribute-mp:

```suggestion
  build:
    uses: crazy-max/.github/.github/workflows/build-distribute-mp.yml@d9a10e2737504a6e253f96e344cef684b0e00cb5
    with:
      target: bin-image
      push: ${{ github.event_name != 'pull_request' && github.repository == 'moby/vpnkit' }}
      cache: true
      meta-image: moby/vpnkit
      meta-tags: |
        type=semver,pattern={{version}}
        type=ref,event=branch
        type=ref,event=pr
        type=sha
      build-platforms: linux/amd64,linux/arm64
    secrets:
      login-username: ${{ secrets.DOCKERHUB_USERNAME }}
      login-password: ${{ secrets.DOCKERHUB_TOKEN }}
```

Reminds me I need to update https://github.com/moby/moby/pull/49474

---

That won't allow to choose a correct gha runner though?

https://github.com/moby/vpnkit/blob/3d4f258e7514b9cc878639f724cbb0caffa8fd98/.github/workflows/bin-image.yml#L70

It's kinda important here because otherwise the build takes ~30 minutes instead of ~5 😅

---

> Ah, actually it does handle it too (`runner` defaults to `auto`), nice!

Yes indeed `auto` choose based on target platform

---

Updated to use your workflow, I did set `cache` to `false` though.

---

> Looks like the warmup stage only builds one platform: https://github.com/docker/vvoland-gha-tests/actions/runs/14330651343/job/40165666013
> 
> Effectively this turns a parallel build into sequential as the linux/arm64 doesn't run until the warmup stage finishes (which caches the linux/amd64).
> 
> Is there a way to disable the warmup without disabling the gha cache completely?

I can add an option for that, will take a look